File: leetcode/0003-longest-substring-without-repeating-characters/solution.rb
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
    window = Hash.new(0)
    longest = 0
    left = 0

    s.length.times do |right|
        window[s[right]] += 1

        while window[s[right]] > 1
            window[s[left]] -= 1
            left += 1
        end

        longest = [longest, right - left + 1].max
    end

    longest
end

File: leetcode/0003-longest-substring-without-repeating-characters/solution.ts
function lengthOfLongestSubstring(s: string): number {
    let charToNextIndex = new Map<string, number>();
    let maxLen = 0;
    let left = 0;
    for (let right = 0; right < s.length; right++) {
        if (charToNextIndex.has(s[right])) {
            left = Math.max(charToNextIndex.get(s[right])!, left);
        }
        maxLen = Math.max(maxLen, right - left + 1);
        charToNextIndex.set(s[right], right + 1);
    }
    return maxLen;
}
